rust/ql/lib/codeql/rust/security/InsecureCookieExtensions.qll

```diff
@@ -0,0 +1,92 @@
+private import codeql.rust.controlflow.BasicBlocks
+
+/**
+ * Provides default sources, sinks and barriers for detecting insecure
+ * cookie vulnerabilities, as well as extension points for adding your own.
+ */
+module InsecureCookie {
+  /**
+   * A data flow source for insecure cookie vulnerabilities.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for insecure cookie vulnerabilities.
+   */
+  abstract class Sink extends QuerySink::Range {
+    override string getSinkType() { result = "InsecureCookie" }
+  }
+
+  /**
+   * A barrier for insecure cookie vulnerabilities.
+   */
+  abstract class Barrier extends DataFlow::Node { }
+
+  /**
+   * A source for insecure cookie vulnerabilities from model data.
+   */
+  private class ModelsAsDataSource extends Source {
+    ModelsAsDataSource() { sourceNode(this, "cookie-create") }
+  }
+
+  /**
+   * A sink for insecure cookie vulnerabilities from model data.
+   */
+  private class ModelsAsDataSink extends Sink {
+    ModelsAsDataSink() { sinkNode(this, "cookie-use") }
+  }
+
+  /**
+   * Holds if cookie attribute `attrib` (`secure` or `partitioned`) is set to `value` (`true` or `false`) at `node`.
+   * A value that cannot be determined is treated as `false`.
+   *
+   * This references models-as-data optional barrier nodes, for example `OptionalBarrier[cookie-secure-arg0]`.
+   */
+  predicate cookieSetNode(DataFlow::Node node, string attrib, boolean value) {
+    exists(
+      FlowSummaryNode summaryNode, string barrierName, CallExprBase ce, int arg,
+      DataFlow::Node argNode
+    |
+      // decode a `cookie-`... optional barrier
+      DataflowImpl::optionalBarrier(summaryNode, barrierName) and
+      attrib = barrierName.regexpCapture("cookie-(secure|partitioned)-arg([0-9]+)", 1) and
+      arg = barrierName.regexpCapture("cookie-(secure|partitioned)-arg([0-9]+)", 2).toInt() and
+      // find a call and arg referenced by this optional barrier
+      ce.getStaticTarget() = summaryNode.getSummarizedCallable() and
+      ce.getArg(arg) = argNode.asExpr().getExpr() and
+      // check if the argument is always `true`
+      (
+        if
+          forex(DataFlow::Node argSourceNode, BooleanLiteralExpr argSourceValue |
+            DataFlow::localFlow(argSourceNode, argNode) and
+            argSourceValue = argSourceNode.asExpr().getExpr() |
+            argSourceValue.getTextValue() = "true"
+          )
+        then value = true // `true` flows to here
+        else value = false // `false`, unknown, or multiple values
+      ) and
+      // and find the node where this happens
+      (
+        node.asExpr().getExpr() = ce.(MethodCallExpr).getReceiver() // e.g. `a` in `a.set_secure(true)`
+        or
+        exists(BasicBlock bb, int i |
+          // associated SSA node
+          node.(SsaNode).asDefinition().definesAt(_, bb, i) and
+          ce.(MethodCallExpr).getReceiver() = bb.getNode(i).getAstNode()
+        )
```

Comment:
These nodes (which function as barriers) are essentially duplicated at the corresponding SSA dataflow nodes, this shouldn't be necessary but it currently is necessary for the state-setting calls (e.g. test line 61) to work properly.